cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -21,57 +21,61 @@ import (
 
 	"github.com/golang/glog"
 	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	kubeinformers "k8s.io/client-go/informers"
 	"k8s.io/client-go/tools/cache"
 	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
-	informers "sigs.k8s.io/cluster-api/pkg/client/informers_generated/externalversions"
+	clusterinformers "sigs.k8s.io/cluster-api/pkg/client/informers_generated/externalversions"
 	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/client/informers_generated/externalversions/cluster/v1alpha1"
 )
 
 const (
-	nodeNameIndexKey = "clusterapi-nodeNameIndex"
+	nodeProviderIDIndex = "clusterapi-nodeProviderIDIndex"
 )
 
-// machineController watches for Machines and MachineSets as they are
-// added, updated and deleted on the cluster.
+// machineController watches for Nodes, Machines and MachineSets as
+// they are added, updated and deleted on the cluster. Additionally,
+// it adds indices to the node informers to satisfy lookup by
+// node.Spec.ProviderID.
 type machineController struct {
-	informerFactory    informers.SharedInformerFactory
-	machineInformer    clusterv1alpha1.MachineInformer
-	machineSetInformer clusterv1alpha1.MachineSetInformer
+	clusterInformerFactory clusterinformers.SharedInformerFactory
+	kubeInformerFactory    kubeinformers.SharedInformerFactory
+	machineInformer        clusterv1alpha1.MachineInformer
+	machineSetInformer     clusterv1alpha1.MachineSetInformer
+	nodeInformer           cache.SharedIndexInformer
 }
 
-func indexMachineByNodeName(obj interface{}) ([]string, error) {
-	machine, ok := obj.(*v1alpha1.Machine)
-	if !ok {
-		return []string{}, nil
-	}
-
-	if machine.Status.NodeRef == nil || machine.Status.NodeRef.Kind != "Node" {
-		return []string{}, nil
+func indexNodeByNodeProviderID(obj interface{}) ([]string, error) {
+	if node, ok := obj.(*apiv1.Node); ok {
+		return []string{node.Spec.ProviderID}, nil
 	}
-
-	glog.V(4).Infof("machine %s/%s is node %q", machine.Namespace, machine.Name, machine.Status.NodeRef.Name)
-
-	return []string{machine.Status.NodeRef.Name}, nil
+	return []string{}, nil
 }
 
-func (c *machineController) findMachine(node *apiv1.Node) (*v1alpha1.Machine, error) {
-	machineIndexer := c.machineInformer.Informer().GetIndexer()
-	objs, err := machineIndexer.ByIndex(nodeNameIndexKey, node.Name)
+func (c *machineController) findMachine(id string) (*v1alpha1.Machine, error) {
+	store := c.machineInformer.Informer().GetStore()
+	item, exists, err := store.GetByKey(id)
```

Comment:
Given the `store` is used only here, it's more transparent to do `c.machineInformer.Informer().GetStore().GetByKey(id)` here.

Reply:
Done.